File: KKY_Tool_Revit_2019-2023/Exports/FamilyLinkAuditExport.vb
Option Explicit On
Option Strict On

Imports System.Data
Imports System.IO
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports KKY_Tool_Revit.Infrastructure
Imports KKY_Tool_Revit.Services

Namespace Exports

    Public Module FamilyLinkAuditExport

        Public ReadOnly Schema As String() = {
            "FileName",
            "HostFamilyName",
            "HostFamilyCategory",
            "NestedFamilyName",
            "NestedTypeName",
            "NestedInstanceId",
            "NestedPath",
            "NestingLevel",
            "NestedCategory",
            "NestedParamName",
            "TargetParamName",
            "ExpectedGuid",
            "FoundScope",
            "NestedParamGuid",
            "NestedParamDataType",
            "AssocHostParamName",
            "HostParamIsShared",
            "Issue",
            "Notes"
        }

        Public Function Export(rows As IEnumerable(Of FamilyLinkAuditRow),
                               Optional fastExport As Boolean = True,
                               Optional autoFit As Boolean = False,
                               Optional progressChannel As String = Nothing) As String
            If rows Is Nothing Then Return String.Empty
            Dim table As DataTable = ToDataTable(rows)
            ' Global.KKY_Tool_Revit.Infrastructure.ResultTableFilter.KeepOnlyIssues("familylink", table)
            ExcelCore.EnsureMessageRow(table, "오류가 없습니다.")
            If Not ValidateSchema(table) Then
                Throw New InvalidOperationException("스키마 검증 실패: 컬럼 순서/헤더가 규격과 다릅니다.")
            End If

            Dim defaultName As String = $"FamilyLinkAudit_{DateTime.Now:yyyyMMdd_HHmm}.xlsx"
            Using dlg As New SaveFileDialog()
                dlg.Filter = "Excel Workbook (*.xlsx)|*.xlsx"
                dlg.FileName = defaultName
                dlg.AddExtension = True
                dlg.DefaultExt = "xlsx"
                dlg.OverwritePrompt = True
                dlg.RestoreDirectory = True

                If dlg.ShowDialog() <> DialogResult.OK Then Return String.Empty

                Dim filePath As String = dlg.FileName
                Dim ext As String = Path.GetExtension(filePath).ToLowerInvariant()
                If Not String.Equals(ext, ".xlsx", StringComparison.OrdinalIgnoreCase) Then
                    filePath = Path.ChangeExtension(filePath, ".xlsx")
                End If

                Dim doAutoFit As Boolean = (Not fastExport) AndAlso autoFit
                Dim excelMode As String = If(fastExport, "fast", "normal")
                ReportProgress(progressChannel, "EXCEL_INIT", "엑셀 저장 준비 중...", 0, Math.Max(1, table.Rows.Count), 0.0R, True)
                ExcelCore.SaveXlsx(filePath, "FamilyLinkAudit", table, doAutoFit, progressKey:=progressChannel, exportKind:="familylink")
                ReportProgress(progressChannel, "EXCEL_SAVE", "엑셀 파일 저장 중...", Math.Max(1, table.Rows.Count), Math.Max(1, table.Rows.Count), 0.95R, True)
                ExcelExportStyleRegistry.ApplyStylesForKey("familylink", filePath, autoFit:=doAutoFit, excelMode:=excelMode)
                If doAutoFit Then
                    ReportProgress(progressChannel, "AUTOFIT", "열 너비 자동 조정 중...", 1, 1, 1.0R, True)
                End If
                ReportProgress(progressChannel, "DONE", "내보내기 완료", 1, 1, 1.0R, True)
                Return filePath
            End Using
        End Function

        Private Sub ReportProgress(channel As String,
                                   phase As String,
                                   message As String,
                                   current As Integer,
                                   total As Integer,
                                   Optional percentOverride As Double? = Nothing,
                                   Optional force As Boolean = False)
            If String.IsNullOrWhiteSpace(channel) Then Return
            Global.KKY_Tool_Revit.UI.Hub.ExcelProgressReporter.Report(channel, phase, message, current, total, percentOverride, force)
        End Sub

        Public Function ToDataTable(rows As IEnumerable(Of FamilyLinkAuditRow)) As DataTable
            Return BuildTable(rows)
        End Function

        Private Function BuildTable(rows As IEnumerable(Of FamilyLinkAuditRow)) As DataTable
            Dim dt As New DataTable("FamilyLinkAudit")
            For Each h In Schema
                dt.Columns.Add(h)
            Next

            For Each r In rows
                Dim dr = dt.NewRow()
                dr("FileName") = SafeStr(r.FileName)
                dr("HostFamilyName") = SafeStr(r.HostFamilyName)
                dr("HostFamilyCategory") = SafeStr(r.HostFamilyCategory)
                dr("NestedFamilyName") = SafeStr(r.NestedFamilyName)
                dr("NestedTypeName") = SafeStr(r.NestedTypeName)
                dr("NestedInstanceId") = SafeStr(r.NestedInstanceId)
                dr("NestedPath") = SafeStr(r.NestedPath)
                dr("NestingLevel") = SafeStr(r.NestingLevel)
                dr("NestedCategory") = SafeStr(r.NestedCategory)
                dr("NestedParamName") = SafeStr(r.NestedParamName)
                dr("TargetParamName") = SafeStr(r.TargetParamName)
                dr("ExpectedGuid") = SafeStr(r.ExpectedGuid)
                dr("FoundScope") = SafeStr(r.FoundScope)
                dr("NestedParamGuid") = SafeStr(r.NestedParamGuid)
                dr("NestedParamDataType") = SafeStr(r.NestedParamDataType)
                dr("AssocHostParamName") = SafeStr(r.AssocHostParamName)
                dr("HostParamIsShared") = SafeStr(r.HostParamIsShared)
                dr("Issue") = SafeStr(r.Issue)
                dr("Notes") = SafeStr(r.Notes)
                dt.Rows.Add(dr)
            Next

            Return dt
        End Function

        Private Function ValidateSchema(table As DataTable) As Boolean
            If table Is Nothing Then Return False
            If table.Columns.Count <> Schema.Length Then Return False
            For i As Integer = 0 To Schema.Length - 1
                If Not String.Equals(table.Columns(i).ColumnName, Schema(i), StringComparison.Ordinal) Then
                    Return False
                End If
            Next
            Return True
        End Function

        Private Function SafeStr(s As String) As String
            Return If(s, "")
        End Function

    End Module

End Namespace

File: KKY_Tool_Revit_2019-2023/Exports/FamilyLinkAuditExport.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using KKY_Tool_Revit.Infrastructure;
using KKY_Tool_Revit.Services;

namespace KKY_Tool_Revit.Exports
{
    public static class FamilyLinkAuditExport
    {
        public static readonly string[] Schema =
        {
            "FileName",
            "HostFamilyName",
            "HostFamilyCategory",
            "NestedFamilyName",
            "NestedTypeName",
            "NestedCategory",
            "NestedParamName",
            "TargetParamName",
            "ExpectedGuid",
            "FoundScope",
            "NestedParamGuid",
            "NestedParamDataType",
            "AssocHostParamName",
            "HostParamIsShared",
            "Issue",
            "Notes"
        };

        public static string Export(IEnumerable<FamilyLinkAuditRow> rows, bool fastExport = true, bool autoFit = false)
        {
            if (rows == null) return string.Empty;
            var table = ToDataTable(rows);
            global::KKY_Tool_Revit.Infrastructure.ResultTableFilter.KeepOnlyIssues("familylink", table);
            ExcelCore.EnsureMessageRow(table, "오류가 없습니다.");
            if (!ValidateSchema(table))
            {
                throw new InvalidOperationException("스키마 검증 실패: 컬럼 순서/헤더가 규격과 다릅니다.");
            }

            var defaultName = $"FamilyLinkAudit_{DateTime.Now:yyyyMMdd_HHmm}.xlsx";
            using (var dlg = new SaveFileDialog())
            {
                dlg.Filter = "Excel Workbook (*.xlsx)|*.xlsx|CSV (*.csv)|*.csv";
                dlg.FileName = defaultName;
                dlg.AddExtension = true;
                dlg.DefaultExt = "xlsx";
                dlg.OverwritePrompt = true;
                dlg.RestoreDirectory = true;

                if (dlg.ShowDialog() != DialogResult.OK) return string.Empty;

                var filePath = dlg.FileName;
                var ext = Path.GetExtension(filePath).ToLowerInvariant();
                if (string.IsNullOrWhiteSpace(ext))
                {
                    ext = dlg.FilterIndex == 2 ? ".csv" : ".xlsx";
                    filePath += ext;
                }

                if (ext == ".csv")
                {
                    SaveCsv(filePath, table);
                }
                else
                {
                    var doAutoFit = !fastExport && autoFit;
                    var excelMode = fastExport ? "fast" : "normal";
                    ExcelCore.SaveXlsx(filePath, "FamilyLinkAudit", table, doAutoFit);
                    ExcelExportStyleRegistry.ApplyStylesForKey("familylink", filePath, autoFit: doAutoFit, excelMode: excelMode);
                }

                return filePath;
            }
        }

        public static DataTable ToDataTable(IEnumerable<FamilyLinkAuditRow> rows)
        {
            return BuildTable(rows);
        }

        private static DataTable BuildTable(IEnumerable<FamilyLinkAuditRow> rows)
        {
            var dt = new DataTable("FamilyLinkAudit");
            foreach (var h in Schema)
            {
                dt.Columns.Add(h);
            }

            foreach (var r in rows)
            {
                var dr = dt.NewRow();
                dr["FileName"] = SafeStr(r.FileName);
                dr["HostFamilyName"] = SafeStr(r.HostFamilyName);
                dr["HostFamilyCategory"] = SafeStr(r.HostFamilyCategory);
                dr["NestedFamilyName"] = SafeStr(r.NestedFamilyName);
                dr["NestedTypeName"] = SafeStr(r.NestedTypeName);
                dr["NestedCategory"] = SafeStr(r.NestedCategory);
                dr["NestedParamName"] = SafeStr(r.NestedParamName);
                dr["TargetParamName"] = SafeStr(r.TargetParamName);
                dr["ExpectedGuid"] = SafeStr(r.ExpectedGuid);
                dr["FoundScope"] = SafeStr(r.FoundScope);
                dr["NestedParamGuid"] = SafeStr(r.NestedParamGuid);
                dr["NestedParamDataType"] = SafeStr(r.NestedParamDataType);
                dr["AssocHostParamName"] = SafeStr(r.AssocHostParamName);
                dr["HostParamIsShared"] = SafeStr(r.HostParamIsShared);
                dr["Issue"] = SafeStr(r.Issue);
                dr["Notes"] = SafeStr(r.Notes);
                dt.Rows.Add(dr);
            }

            return dt;
        }

        private static bool ValidateSchema(DataTable table)
        {
            if (table == null) return false;
            if (table.Columns.Count != Schema.Length) return false;

            for (var i = 0; i < Schema.Length; i++)
            {
                if (!string.Equals(table.Columns[i].ColumnName, Schema[i], StringComparison.Ordinal))
                {
                    return false;
                }
            }

            return true;
        }

        private static void SaveCsv(string filePath, DataTable table)
        {
            using (var fs = new FileStream(filePath, FileMode.Create, FileAccess.Write, FileShare.Read))
            using (var sw = new StreamWriter(fs, new UTF8Encoding(true)))
            {
                sw.WriteLine(string.Join(",", Schema.Select(CsvEscape)));
                foreach (DataRow row in table.Rows)
                {
                    var cols = new List<string>();
                    foreach (var h in Schema)
                    {
                        cols.Add(CsvEscape((row[h] ?? string.Empty).ToString()));
                    }

                    sw.WriteLine(string.Join(",", cols));
                }
            }
        }

        private static string CsvEscape(string s)
        {
            if (s == null) s = string.Empty;
            const string quote = "\"";
            var needsQuotes = s.Contains(',') || s.Contains(quote) || s.Contains("\r") || s.Contains("\n");
            s = s.Replace(quote, quote + quote);
            return needsQuotes ? quote + s + quote : s;
        }

        private static string SafeStr(string s)
        {
            return s ?? string.Empty;
        }
    }
}
